File: plugins/capacitor-plugin-local-llm/ios/Sources/LocalLLMPlugin/LocalLLMPlugin.swift
import Foundation
import Capacitor

@objc(LocalLLMPlugin)
public class LocalLLMPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "LocalLLMPlugin"
    public let jsName = "LocalLLM"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "downloadModel", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "deleteModel", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getModelStatus", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getAvailableDiskSpace", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "loadModel", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "unloadModel", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "generate", returnType: CAPPluginReturnPromise),
    ]

    private let engine = LlamaEngine()
    private let manager = ModelManager.shared

    @objc func downloadModel(_ call: CAPPluginCall) {
        guard let modelName = call.getString("modelName") else {
            call.reject("modelName is required")
            return
        }
        Task {
            do {
                try await manager.downloadModel(modelName) { [weak self] progress, downloaded, total in
                    self?.notifyListeners("downloadProgress", data: [
                        "progress": progress,
                        "bytesDownloaded": downloaded,
                        "totalBytes": total
                    ])
                }
                call.resolve(["success": true])
            } catch {
                call.reject("Download failed: \(error.localizedDescription)")
            }
        }
    }

    @objc func deleteModel(_ call: CAPPluginCall) {
        guard let modelName = call.getString("modelName") else {
            call.reject("modelName is required")
            return
        }
        do {
            engine.unload()
            try manager.deleteModel(modelName)
            call.resolve()
        } catch {
            call.reject("Delete failed: \(error.localizedDescription)")
        }
    }

    @objc func getModelStatus(_ call: CAPPluginCall) {
        guard let modelName = call.getString("modelName") else {
            call.reject("modelName is required")
            return
        }
        call.resolve([
            "downloaded": manager.isDownloaded(modelName),
            "sizeBytes": manager.modelSize(modelName),
            "path": manager.modelPath(modelName).path
        ])
    }

    @objc func getAvailableDiskSpace(_ call: CAPPluginCall) {
        call.resolve(["bytes": manager.availableDiskSpace()])
    }

    @objc func loadModel(_ call: CAPPluginCall) {
        guard let modelName = call.getString("modelName") else {
            call.reject("modelName is required")
            return
        }
        let contextSize = Int32(call.getInt("contextSize") ?? 4096)
        let gpuLayers = Int32(call.getInt("gpuLayers") ?? 99)
        let path = manager.modelPath(modelName).path

        Task.detached { [engine] in
            do {
                try engine.load(modelPath: path, contextSize: contextSize, gpuLayers: gpuLayers)
                call.resolve(["loaded": true])
            } catch {
                call.reject("Load failed: \(error.localizedDescription)")
            }
        }
    }

    @objc func unloadModel(_ call: CAPPluginCall) {
        engine.unload()
        call.resolve()
    }

    @objc func generate(_ call: CAPPluginCall) {
        guard let prompt = call.getString("prompt") else {
            call.reject("prompt is required")
            return
        }
        let systemPrompt = call.getString("systemPrompt")
        let maxTokens = call.getInt("maxTokens") ?? 2048
        let temperature = call.getFloat("temperature") ?? 0.3
        let stopSequences = call.getArray("stopSequences", String.self) ?? []

        Task.detached { [engine] in
            do {
                let result = try engine.generate(
                    prompt: prompt,
                    systemPrompt: systemPrompt,
                    maxTokens: maxTokens,
                    temperature: temperature,
                    stopSequences: stopSequences
                )
                call.resolve([
                    "content": result.content,
                    "tokensUsed": result.tokensUsed
                ])
            } catch {
                call.reject("Generation failed: \(error.localizedDescription)")
            }
        }
    }
}

File: plugins/capacitor-plugin-local-llm/android/src/main/java/com/neurocanvas/localllm/LocalLLMPlugin.kt
package com.neurocanvas.localllm

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import kotlinx.coroutines.*

@CapacitorPlugin(name = "LocalLLM")
class LocalLLMPlugin : Plugin() {
    private val engine = LlamaEngine()
    private lateinit var manager: ModelManager
    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    override fun load() {
        manager = ModelManager(context)
    }

    @PluginMethod
    fun downloadModel(call: PluginCall) {
        val modelName = call.getString("modelName") ?: return call.reject("modelName is required")

        scope.launch {
            try {
                manager.downloadModel(modelName) { progress, downloaded, total ->
                    val data = JSObject().apply {
                        put("progress", progress)
                        put("bytesDownloaded", downloaded)
                        put("totalBytes", total)
                    }
                    notifyListeners("downloadProgress", data)
                }
                call.resolve(JSObject().apply { put("success", true) })
            } catch (e: Exception) {
                call.reject("Download failed: ${e.message}")
            }
        }
    }

    @PluginMethod
    fun deleteModel(call: PluginCall) {
        val modelName = call.getString("modelName") ?: return call.reject("modelName is required")
        engine.unload()
        manager.deleteModel(modelName)
        call.resolve()
    }

    @PluginMethod
    fun getModelStatus(call: PluginCall) {
        val modelName = call.getString("modelName") ?: return call.reject("modelName is required")
        call.resolve(JSObject().apply {
            put("downloaded", manager.isDownloaded(modelName))
            put("sizeBytes", manager.modelSize(modelName))
            put("path", manager.modelPath(modelName).absolutePath)
        })
    }

    @PluginMethod
    fun getAvailableDiskSpace(call: PluginCall) {
        call.resolve(JSObject().apply { put("bytes", manager.availableDiskSpace()) })
    }

    @PluginMethod
    fun loadModel(call: PluginCall) {
        val modelName = call.getString("modelName") ?: return call.reject("modelName is required")
        val contextSize = call.getInt("contextSize") ?: 4096
        val gpuLayers = call.getInt("gpuLayers") ?: 0

        scope.launch(Dispatchers.IO) {
            val path = manager.modelPath(modelName).absolutePath
            val loaded = engine.load(path, contextSize, gpuLayers)
            withContext(Dispatchers.Main) {
                call.resolve(JSObject().apply { put("loaded", loaded) })
            }
        }
    }

    @PluginMethod
    fun unloadModel(call: PluginCall) {
        engine.unload()
        call.resolve()
    }

    @PluginMethod
    fun generate(call: PluginCall) {
        val prompt = call.getString("prompt") ?: return call.reject("prompt is required")
        val systemPrompt = call.getString("systemPrompt")
        val maxTokens = call.getInt("maxTokens") ?: 2048
        val stopSequences = call.getArray("stopSequences")?.toList<String>() ?: emptyList()

        scope.launch(Dispatchers.IO) {
            try {
                val (content, tokensUsed) = engine.generate(prompt, systemPrompt, maxTokens, stopSequences)
                withContext(Dispatchers.Main) {
                    call.resolve(JSObject().apply {
                        put("content", content)
                        put("tokensUsed", tokensUsed)
                    })
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    call.reject("Generation failed: ${e.message}")
                }
            }
        }
    }

    override fun handleOnDestroy() {
        engine.unload()
        scope.cancel()
    }
}
